viewing_party/party.py

```diff
@@ -1,23 +1,175 @@
+            del user_data["watchlist"][i]
+
+    return user_data    
 
 # -----------------------------------------
 # ------------- WAVE 2 --------------------
 # -----------------------------------------
 
+def get_watched_avg_rating(user_data):
+    if not user_data["watched"]:
+        return 0.0
+    
+    count = 0
+    sum_ratings = 0
+
+    for movie in user_data["watched"]:
+        sum_ratings += movie["rating"]
+        count += 1
+
+    avg_rating = sum_ratings/count
+
+    return avg_rating
+
+def get_most_watched_genre(user_data):
+    genres_count = {}
+    
+    # make a dictionary where genre is key and 
+    # how many times genre watched is value
+    if not user_data["watched"]:
+        return None
+    else:
+        for movie in user_data["watched"]:
+            genres_count[movie["genre"]] = genres_count.get(movie["genre"], 0) + 1
+    
+    most_watched_genre = None
+    times = 0
+    # iterate through genres_count dict
+    # find genre with highest count
+    for genre, count in genres_count.items():
+        if count > times:
+            most_watched_genre = genre
+            times = count
+    
+    return most_watched_genre
 
 # -----------------------------------------
 # ------------- WAVE 3 --------------------
 # -----------------------------------------
 
+def get_unique_watched(user_data):
+    unique_watched = []
+    set_user_mov_titles = set()
+    set_mov_friends_titles = set()
+
+    # create set of friends movies titles
+    for fr_watched in user_data["friends"]:
+        for mov in fr_watched["watched"]:
+            set_mov_friends_titles.add(mov["title"])
+        
+    # create set of user movies titles
+    for movie in user_data["watched"]:
+        set_user_mov_titles.add(movie["title"])
+
+    # compare sets to determine only those that only user have seen
+    unique_titles = set_user_mov_titles.difference(set_mov_friends_titles)
+    
+    # add those unique movies to a list of movies based on title
+    for movie in user_data["watched"]:
+        if movie["title"] in unique_titles:
+            unique_watched.append(movie)
+
+    return unique_watched
+
+def get_friends_unique_watched(user_data):
+    unique_watched = []
+    set_user_mov_titles = set()
+    set_mov_friends_titles = set()
+
+    # create set of friends movies titles
+    for fr_watched in user_data["friends"]:
+        for mov in fr_watched["watched"]:
+            set_mov_friends_titles.add(mov["title"])
+
+    # create set of user movies titles
+    for movie in user_data["watched"]:
+        set_user_mov_titles.add(movie["title"])
+
+    # compare sets to determine only those that user have never seen
+    unique_titles = set_mov_friends_titles.difference(set_user_mov_titles)
+
+    # add those movies to the list based on titles
+    for fr_watched in user_data["friends"]:
+        for mov in fr_watched["watched"]:
+            if mov["title"] in unique_titles and mov not in unique_watched:

```

Comment:
Great idea to check and see if the movie is already in `unique_watched` before adding it to the result list!